pancake/proofs/loop_to_wordProofScript.sml

```diff
@@ -1485,4 +1485,532 @@ Proof
+  drule loop_to_word_comp_no_mt>>gs[]>>
+  qpat_x_assum ‘(p1, _) = _’ kall_tac>>
+  drule loop_to_word_comp_no_mt>>gs[]
+QED
+
+Theorem loop_to_word_compile_no_mt:
+  wprog = compile_prog pan_prog ⇒
+  EVERY no_mt (MAP (SND o SND) wprog)
+Proof
+  qid_spec_tac ‘wprog’>>
+  Induct_on ‘pan_prog’>>
+  gs[compile_def, compile_prog_def]>>
+  strip_tac>>pairarg_tac>>gs[loop_to_word_comp_func_no_mt]
+QED
+
+Theorem loop_compile_no_mt_code:
+  compile prog = prog' ⇒
+  no_mt_code (fromAList prog')
+Proof
+  disch_then (assume_tac o GSYM)>>
+  gs[compile_def]>>
+  drule loop_to_word_compile_no_mt>>
+  rw[wordPropsTheory.no_mt_code_def]>>
+  gs[lookup_fromAList, EVERY_MEM, MEM_MAP]>>
+  drule ALOOKUP_MEM>>strip_tac>>
+  first_x_assum (qspec_then ‘p’ assume_tac)>>
+  res_tac>>gs[]
+QED
+
+(*** loop_to_word good_handlers ***)
+
+Theorem comp_l_invariant:
+  ∀ctxt prog l prog' l'. comp ctxt prog l = (prog',l') ⇒ FST l' = FST l
+Proof
+  ho_match_mp_tac comp_ind >>
+  rw[comp_def] >>
+  gvs[ELIM_UNCURRY,PULL_FORALL,AllCaseEqs()] >> metis_tac[PAIR]
+QED
+
+Theorem good_handlers_comp:
+  ∀ctxt prog l. good_handlers (FST l) (FST (comp ctxt prog l))
+Proof
+  ho_match_mp_tac comp_ind >>
+  rw[wordPropsTheory.good_handlers_def,
+     comp_def] >>
+  gvs[ELIM_UNCURRY] >>
+  rpt(PURE_TOP_CASE_TAC >> gvs[]) >>
+  metis_tac[PAIR,FST,SND,comp_l_invariant]
+QED
+
+Theorem loop_to_word_good_handlers:
+  (compile_prog prog) = prog' ⇒
+  EVERY (λ(n,m,pp). good_handlers n pp) prog'
+Proof
+  fs[compile_def,
+     compile_prog_def,
+     comp_func_def]>>
+  rw[]>>
+  fs[EVERY_MEM,MEM_MAP,PULL_EXISTS]>>
+  PairCases >>
+  rw[] >>
+  pop_assum kall_tac >>
+  rename1 ‘comp ctxt prog’ >>
+  rename1 ‘(n,m)’ >>
+  metis_tac[PAIR,FST,SND,good_handlers_comp]
+QED
+
+(**** lab_pres for loop_to_word ****)
+
+Theorem loop_to_word_comp_SND_LE:
+  ∀ctxt prog l p r.
+    comp ctxt prog l = (p,r) ⇒
+    SND l ≤ SND r
+Proof
+  ho_match_mp_tac loop_to_wordTheory.comp_ind>>
+  rw[loop_to_wordTheory.comp_def]>>
+  rpt (FULL_CASE_TAC>>gs[])>>TRY (rveq>>gs[])>>
+  pairarg_tac>>gs[]>>pairarg_tac>>gs[]>>
+  rveq>>gs[]
+QED
+
+Theorem loop_to_word_comp_extract_labels_len:
+  ∀ctxt prog l p r.
+    comp ctxt prog l = (p,r) ⇒
+    LENGTH (extract_labels p) = SND r - SND l
+Proof
+  ho_match_mp_tac loop_to_wordTheory.comp_ind>>
+  rw[loop_to_wordTheory.comp_def]>>
+  gs[wordPropsTheory.extract_labels_def]
+  >- (
+  pairarg_tac>>gs[]>>
+  pairarg_tac>>gs[]>>
+  rveq>>gs[wordPropsTheory.extract_labels_def]>>
+  drule loop_to_word_comp_SND_LE>>strip_tac>>
+  qpat_x_assum ‘_= (_, l')’ assume_tac>>
+  drule loop_to_word_comp_SND_LE>>strip_tac>>
+  gs[])
+  >- (
+  pairarg_tac>>gs[]>>
+  pairarg_tac>>gs[]>>
+  rveq>>gs[wordPropsTheory.extract_labels_def]>>
+  drule loop_to_word_comp_SND_LE>>strip_tac>>
+  qpat_x_assum ‘_= (_, l')’ assume_tac>>
+  drule loop_to_word_comp_SND_LE>>strip_tac>>
+  gs[])>>
+  rpt (FULL_CASE_TAC>>gs[])>>
+  rveq>>gs[wordPropsTheory.extract_labels_def]>>
+  TRY (CASE_TAC>>gs[])>>
```

Comment:
avoid TRY and attempt to avoid CASE_TAC, use AllCaseEqs() instead wherever possible